core/object/script_language.h

```diff
@@ -274,13 +274,34 @@ class ScriptLanguage {
 		String message;
 	};
 
+	enum ScriptOrigin {
+		SCRIPT_ORIGIN_BUILT_IN,
+		SCRIPT_ORIGIN_EDITOR,
+		SCRIPT_ORIGIN_PROJECT,
+	};
+
+	struct ScriptTemplate {
+		String inherit = "Object";
+		String name;
+		String description;
+		String content;
+		bool default_template = false;
+		int id = 0;
+		ScriptOrigin origin = ScriptOrigin::SCRIPT_ORIGIN_BUILT_IN;
+
+		String get_hash() const {
+			return itos(origin) + inherit + name;
+		}
+	};
+
 	void get_core_type_words(List<String> *p_core_type_words) const;
 	virtual void get_reserved_words(List<String> *p_words) const = 0;
 	virtual bool is_control_flow_keyword(String p_string) const = 0;
 	virtual void get_comment_delimiters(List<String> *p_delimiters) const = 0;
 	virtual void get_string_delimiters(List<String> *p_delimiters) const = 0;
-	virtual Ref<Script> get_template(const String &p_class_name, const String &p_base_class_name) const = 0;
-	virtual void make_template(const String &p_class_name, const String &p_base_class_name, Ref<Script> &p_script) {}
+	virtual Ref<Script> make_template(const String &p_template, const String &p_class_name, const String &p_base_class_name) const { return Ref<Script>(); }
+	virtual Ref<Script> get_plugin_template(const String &p_class_name, const String &p_base_class_name) const { return Ref<Script>(); }
+	virtual Vector<ScriptTemplate> get_built_in_template(StringName p_object) { return Vector<ScriptTemplate>(); }
```

Comment:
If the method returns vector of templates, shouldn't it be called `get_built_in_templates()`?